src/ray/common/id.h

```diff
@@ -451,11 +457,12 @@ template <typename T>
 std::string BaseID<T>::Hex() const {
   constexpr char hex[] = "0123456789abcdef";
   const uint8_t *id = Data();
-  std::string result;
+  std::string result(2 * T::Size(), '\0');
+  char *dst = &result.front();
   for (int i = 0; i < T::Size(); i++) {
     unsigned int val = id[i];
-    result.push_back(hex[val >> 4]);
-    result.push_back(hex[val & 0xf]);
+    *dst++ = hex[val >> 4];
+    *dst++ = hex[val & 0xf];
```

Comment:
Nice refine!

But I believe this could be simplified to :
```C++
std::string result(2 * T::Size(), '\0');
for (int i = 0; i < T::Size(); i++) {
    const auto &val = id[i];
    result[i] = hex[val >> 4];
    result[i + 1] = hex[val & 0xf];
}
```

Reply:
> Nice refine!
> 
> But I believe this could be simplified to :
> 
> ```c++
> std::string result(2 * T::Size(), '\0');
> for (int i = 0; i < T::Size(); i++) {
>     const auto &val = id[i];
>     result[i] = hex[val >> 4];
>     result[i + 1] = hex[val & 0xf];
> }
> ```

I think what you wrote above need to be:
```
std::string result(2 * T::Size(), '\0');
for (int i = 0; i < T::Size(); i++) {
    const auto &val = id[i];
    result[2*i] = hex[val >> 4];
    result[2*i + 1] = hex[val & 0xf];
}
```
I do not recommend doing like this.  It is actually the least efficient of the previous methods as the `result[]` invoke the `operator[]` method which is not so fast than `*dst++`.

Think of what I wrote inside this PR as A, the original method as B, , and yours as C. The performance is `A good than B good than C`.